sensor/common/processsignal/enricher.go

```diff
@@ -106,9 +127,12 @@ func (e *enricher) processLoop() {
 
 // scans the cache and enriches indicators that have metadata.
 func (e *enricher) scanAndEnrich(metadata clusterentities.ContainerMetadata) {
-	if wrapInterface, ok := e.lru.Peek(metadata.ContainerID); ok {
-		wrap := wrapInterface.(*containerWrap)
-		for _, indicator := range wrap.processes {
+	if wrapObj, ok := e.lru.Peek(metadata.ContainerID); ok {
+		wrap := wrapObj.(*containerWrap)
+		n := wrap.numProcesses()
```

Comment:
All these issues can be elegantly solved by implementing a
```
func (cw *containerWrap) fetchProcesses() []*storage.ProcessIndicators {
  cw.lock.Lock()
  defer cw.lock.Unlock()

  processes := cw.processes
  cw.processes = nil
  return processes
}
```

method, and then changing the above order of operations to
```
	if wrapObj, ok := e.lru.Peek(metadata.ContainerID); ok {
		e.lru.Remove(metadata.ContainerID)
		processes := wrapObj.(*containerWrap).fetchProcesses()
		// wrapObj is no longer used
```
Because `cw.processes` no longer points to the returned `processes` slice, there can be no concurrent insertions (assuming that all insertions still happen via `addProcess`, which requires an exclusive lock).

While it could be the case that some other goroutine has obtained a reference to `wrapObj` in `(*enricher).Add`, the use of a mutex ensures that the addition of the new process either happens strictly before the call to `e.lru.Remove` in `scanAndEnrich`, or after. If it happens before the call to `e.lru.Remove`, we are guaranteed that this new process will be returned in `fetchProcesses()` (because that is only called after `e.lru.Remove`). If it happens after, it may or may not be returned depending on the interleaving, but in either case, the `wrapObj` will be re-inserted into `e.lru` in `(*enricher).Add` and thus any processes inserted into the (new) `cw.processes` slice after `fetchProcesses` was called will not be lost.

It is possible that we have `containerWrap` instances in the LRU cache with `nil` `cw.processes` slices, but that does not affect correctness.

Reply:
This is a great idea I will switch to this suggestion. Thanks!